#1217 - Player rating graph - show rating on mouse over

Marc-Spector opened the pull request:
Closes #1217

Marc-Spector commented:
<details>
  <summary>Demo (updated)</summary>  
     
![demo](https://user-images.githubusercontent.com/38242291/116887881-79b2c280-ac33-11eb-85fd-aa83c62d72c1.gif)


</details>

Marc-Spector commented:
The PR is ready to review

Sheikah45 commented on src/main/java/com/faforever/client/fx/LineChartWithHover.java:
```diff
@@ -0,0 +1,176 @@
+    super(xAxis, yAxis);
+    chartBackground = (Region) lookup(".chart-plot-background");
+    if (chartBackground == null) {
+      throw new RuntimeException("'Chart background' object was no found. Using LineChartWithHover class is impossible");
```
```suggestion
      throw new RuntimeException("'Chart background' object was not found. Using LineChartWithHover class is impossible");
```

Marc-Spector replied:
You checked the code of old commits. The exception is no longer used and replaced with the `available` flag.

Sheikah45 reviewed (changes requested):
It would be nice to add some tests to the LineChartWithHover

Sheikah45 commented on src/main/java/com/faforever/client/chat/PlayerRatingChartTooltipController.java:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  @Override
+  public void currentXY(Double date, Double rating) {
```
call it setXY

Sheikah45 replied:
Also the XValues are always longs and the rating values are always integers and should be typed as such and use the primitive types

Sheikah45 commented on src/main/java/com/faforever/client/chat/PlayerRatingChartTooltipController.java:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  @Override
+  public void empty() {
```
call it clear

Sheikah45 commented on src/main/java/com/faforever/client/fx/LineChartWithHover.java:
```diff
@@ -0,0 +1,190 @@
+    Tooltip.install(chartBackground, hoverTooltip);
+  }
+
+  private void showCurrentValues(MouseEvent event) {
```
I would prefer setValues

Sheikah45 commented on src/main/java/com/faforever/client/fx/LineChartWithHover.java:
```diff
@@ -0,0 +1,190 @@
+    }
+  }
+
+  private void moveTooltipAndLineToCursor(MouseEvent event) {
```
It appears this is being called twice. Once inside of showCurrentValues and again as a direct mouse event listener. Only call it once

Sheikah45 commented on src/main/java/com/faforever/client/fx/LineChartWithHover.java:
```diff
@@ -0,0 +1,190 @@
+          .forEach((node) -> node.setMouseTransparent(true));
+      chartBackground.setOnMouseMoved((event) -> {
+        showCurrentValues(event);
+        moveTooltipAndLineToCursor(event);
```
This is a potentially redundant call as it is already called in showCurrentValues

Sheikah45 commented on src/main/java/com/faforever/client/fx/LineChartWithHover.java:
```diff
@@ -0,0 +1,190 @@
+
+  public interface LineChartTooltip extends Controller<Node> {
+
+    void currentXY(Double x, Double y);
```
setXY

Sheikah45 commented on src/main/java/com/faforever/client/fx/LineChartWithHover.java:
```diff
@@ -0,0 +1,190 @@
+
+    void currentXY(Double x, Double y);
+
+    void empty();
```
clear

Sheikah45 commented on src/main/java/com/faforever/client/fx/LineChartWithHover.java:
```diff
@@ -0,0 +1,190 @@
+    return chartBackground;
+  }
+
+  public interface LineChartTooltip extends Controller<Node> {
```
This should be moved to its own class

Sheikah45 commented on src/main/java/com/faforever/client/util/TimeService.java:
```diff
@@ -119,7 +119,7 @@ private Locale getCurrentTimeLocale() {
   }
 
-  private Locale getCurrentDateLocale() {
+  public Locale getCurrentDateLocale() {
```
This should not be made public. Conversion should use asDate instead of making its own method.

Sheikah45 commented on src/main/java/com/faforever/client/chat/PlayerRatingChartTooltipController.java:
```diff
@@ -0,0 +1,55 @@
+
+  @Override
+  public void initialize() {
+    dateFormat = DateFormat.getDateInstance(DateFormat.MEDIUM, timeService.getCurrentDateLocale());
```
There is not need to have its own date format should just use timeService.asDate

Sheikah45 commented on src/main/java/com/faforever/client/chat/PlayerRatingChartTooltipController.java:
```diff
@@ -0,0 +1,55 @@
+  @Override
+  public void currentXY(Double date, Double rating) {
+    JavaFxUtil.runLater(() -> {
+      dateLabel.setText(dateFormat.format(TimeUnit.SECONDS.toMillis(date.longValue())));
```
```suggestion
     dateLabel.setText(timeService.asDate(Instant.ofEpochSeconds(date)));
```

Sheikah45 commented on src/main/java/com/faforever/client/fx/LineChartWithHover.java:
```diff
@@ -0,0 +1,190 @@
+      LineTo line1 = lines.get(i);
+      LineTo line2 = lines.get(i + 1);
+      int x1 = (int) Math.floor(line1.getX());
+      int x2 = (int) Math.floor(line2.getX());
```
you can just cast directly to int it has the same effect

Marc-Spector replied:
Hmmm... It is impossible.
![image](https://user-images.githubusercontent.com/38242291/120098309-dffd0900-c13d-11eb-943a-b58fbdbe3336.png)

1-alex98 replied:
(int)

1-alex98 replied:
Your screenshot does not show a cast

Sheikah45 commented on src/main/java/com/faforever/client/fx/LineChartWithHover.java:
```diff
@@ -0,0 +1,190 @@
+      int x1 = (int) Math.floor(line1.getX());
+      int x2 = (int) Math.floor(line2.getX());
+      Number yValue1 = getDisplayedYValue(line1.getY());
+      Number yValue2 = getDisplayedYValue(line2.getY());
```
getDisplayedValue already returns a double and you use them as doubles no need to cast to Number

Sheikah45 replied:
Need a null check here

Sheikah45 commented on src/main/java/com/faforever/client/fx/LineChartWithHover.java:
```diff
@@ -0,0 +1,190 @@
+  }
+
+  private Double getDisplayedXValue(double displayPosition) {
+    return (Double) getXAxis().getValueForDisplay(displayPosition);
```
This casting will only work when the axis is of double objects so this class may only be valid for when X and Y are doubles

Sheikah45 commented on src/main/java/com/faforever/client/fx/LineChartWithHover.java:
```diff
@@ -0,0 +1,190 @@
+  }
+
+  private Double getDisplayedYValue(double displayPosition) {
+    return (Double) getYAxis().getValueForDisplay(displayPosition);
```
same as with X

Sheikah45 reviewed (changes requested):
Apologies for the delay in the review did not realize this was ready.

I found some confusing usage of the generics in the LineChart class where there is direct casting to a double which is not valid in all cases so should either be extended to be really generic or the types should be narrowed to only being doubles.

Also throughout it is better to use the primitive types where possible rather than the boxed types.

IvanPavilionG7 committed:
show rating on mouse over in rating graph

IvanPavilionG7 committed:
remove PlayerRatingChartTooltip interface

Marc-Spector commented:
@Sheikah45 Thanks for the review. Updated the PR.

1-alex98 commented on src/test/java/com/faforever/client/fx/MouseEvents.java:
```diff
@@ -11,4 +11,12 @@ private MouseEvents() {
+   * @param x the x with respect to the node.
+   * @param y the y with respect to the node.
+   */
+  public static MouseEvent generateMouseMoved(int x, int y) {
```
FxRobot class does such things I think

Marc-Spector replied:
Cannot find the method I need. The `moveTo` method is not suitable because he moves the cursor physically. The `generateMouseMove` simply imitates the movement of the cursor to the node for invoking of `mouseMoved` listener.

1-alex98 replied:
ok

1-alex98 commented on src/main/java/com/faforever/client/fx/PlayerRatingChart.java:
```diff
@@ -0,0 +1,187 @@
+      }
+      int distance = x2 - x1;
+      if (distance > 1) {
+        double k = (double) (rating2 - rating1) / distance;
```
Bad variable naming. Maybe `gradient` or what is this k?

1-alex98 commented on src/main/java/com/faforever/client/fx/PlayerRatingChart.java:
```diff
@@ -0,0 +1,187 @@
+    for (int i = 0; i < lines.size() - 1; i++) {
+      LineTo line1 = lines.get(i);
+      LineTo line2 = lines.get(i + 1);
+      int x1 = (int) Math.floor(line1.getX());
+      int x2 = (int) Math.floor(line2.getX());
```
leftXCoordinate rightXCordinate maybe

1-alex98 commented on src/main/java/com/faforever/client/fx/PlayerRatingChart.java:
```diff
@@ -0,0 +1,187 @@
+  }
+
+  private void buildData(List<PathElement> elements) {
+    List<LineTo> lines = elements.stream().filter((e) -> e instanceof LineTo).map((e) -> (LineTo) e).collect(Collectors.toList());
```
e stands for element? or energy 🤔 XD

1-alex98 commented on src/main/java/com/faforever/client/fx/PlayerRatingChart.java:
```diff
@@ -0,0 +1,187 @@
+  }
+
+  @Override
+  protected void seriesChanged(Change<? extends Series> c) {
```
`change`

Sheikah45 commented on src/main/java/com/faforever/client/chat/PlayerRatingChartTooltipController.java:
```diff
@@ -0,0 +1,48 @@
+  public void setDateAndRating(long dateValueInSec, int rating) {
+    JavaFxUtil.runLater(() -> {
+      dateLabel.setText(timeService.asDate(Instant.ofEpochSecond(dateValueInSec)));
+      ratingLabel.setText(Long.toString(Math.round(rating)));
```
rating is already an int why round it?

Marc-Spector replied:
I missed this code when I updated the type of the `rating` parameter from `double` to `int`. Fixed.

Geosearchef replied:
The client code uses `String.valueOf()` in most other locations, so I'd go for that one instead of `Long.toString()`. It's also null safe in case the object in question is null, which doesn't matter for this case though.

Sheikah45 commented on src/main/java/com/faforever/client/fx/PlayerRatingChart.java:
```diff
@@ -0,0 +1,187 @@
+
+  private void putMapRatingValue(int x, double rating) {
+    Double currentRating = ratingMap.get(x);
+    ratingMap.put(x, currentRating != null ? (currentRating + rating) / 2 : rating);
```
look at the merge function of a map. It would make this easier to read

IvanPavilionG7 committed:
make variables naming more clear

IvanPavilionG7 committed:
remove empty method `initialize`

IvanPavilionG7 committed:
move `PlayerRatingChart` class from `com.faforever.client.fx` to
`com.faforever.client.chat`

IvanPavilionG7 committed:
more clean code

IvanPavilionG7 committed:
fix comment

Geosearchef commented on src/main/java/com/faforever/client/fx/PlayerRatingChart.java:
```diff
@@ -0,0 +1,187 @@
+
+  private long getDisplayedDateValue(double displayPosition) {
+    Double value = getXAxis().getValueForDisplay(displayPosition);
+    return value != null ? value.longValue() : Long.MIN_VALUE;
```
```
return Optional.ofNullable(getXAxis().getValueForDisplay(displayPosition))
             .map(ClassNameGoesHere::longValue)
             .else(Long.MIN_VALUE)
```
Just a suggestion, this temp value variable and tertiary op look very verbose, not sure if it's any better though.

Marc-Spector replied:
Done. :)

Sheikah45 commented on src/main/java/com/faforever/client/chat/PlayerRatingChartTooltipController.java:
```diff
@@ -0,0 +1,45 @@
+  public void setDateAndRating(long dateValueInSec, int rating) {
+    JavaFxUtil.runLater(() -> {
+      dateLabel.setText(timeService.asDate(Instant.ofEpochSecond(dateValueInSec)));
+      ratingLabel.setText(Integer.toString(rating));
```
one small change here. i18n has a number method so this would become
`ratingLabel.setText(i18n.number(rating))`

Marc-Spector replied:
Done.

IvanPavilionG7 committed:
more clean code
use `i18n.number` instead of `Integer.toString`
rename test methods:
 from `testSetXYWhenMouseMovingOverLine` to
 `testSetDateAndRatingWhenMouseMovingOverLine` and
 from `testNoXYWhenMouseMovingOverWithoutLine` to
 `testClearDataAndRatingWhenMouseMovingOverWithoutLine`

Sheikah45 commented:
Although checks are failing as i18n is not in the test class instantiation

IvanPavilionG7 committed:
add missed mock i18n service to `PlayerRatingChartTooltipControllerTest`
class
